[RFC] `wildcard` field migration - stage 3 (#1530)
elastic/ecs: rfcs/text/0001-wildcard-data-type.md

### Identified Wildcard Fields

> It will require changing the field's defined schema `type` from `keyword` to `wildcard` for a field to use the wildcard data type.

Reviewer: FWIW I find the original language more clear

}
```

**Note**: the existing `text` data type multi-field will remain if there is a need to support tokenized searches.

## Usage

<!--
Stage 1: Describe at a high-level how these field changes will be used in practice. Real world examples are encouraged. The goal here is to understand how people would leverage these fields to gain insights or solve problems. ~1-3 paragraphs.
-->

Wildcard is well-suited for cases requiring partial matching of string values across long unstructured or semi-structured fields. Often, machine-generated events, such as logs, metrics, and traces, aren't well suited for the analysis applied on `text` fields. Using `keyword` allows for exact value searching and introduces filtering, sorting, and aggregations. Keyword fields also support regex and wildcard queries, however, the field's search performance can vary, depending on the query (for example, leading wildcard) and the cardinality of the data.

> Often ECS approaches semi-structured fields by breaking their values down into structured ones. These structured fields then enable better use of keyword fields' characteristics. For example, a URL can break down into its constituent parts: scheme, domain, port, path, etc., and those parts can, in turn, map to unique fields. However, not all fields are structured enough for this approach.

Reviewer: Another way to put it: `wildcard` works well if you don't know the full structure or can't tokenize a string to accommodate all use cases. Wildcard gives more flexibility at run time to make decisions, instead of solving everything at index time.

Let me know if you need me to elaborate further.

| Supports normalizers in mappings | Y | N |
| Indexing speeds | Fast | Slower (see *6) |

1. Somewhat slower as doc values retrieved from compressed blocks of 32.
2. Somewhat slower because approximate matches with n-grams need verification.

> 3. Keyword field visits every unique value only once, but wildcard field assesses every utterance of values.

Reviewer: what's an utterance?

Author: The table and its footnotes are taken from the [wildcard intro blog post](https://www.elastic.co/blog/find-strings-within-strings-faster-with-the-new-elasticsearch-wildcard-field), and I believe the point here is that for a wildcard field, Elasticsearch will need to traverse all the 3-character sequences indexed from the original string.

4. If "allow expensive queries" is enabled.
5. Depends on common prefixes - keyword fields have common-prefix-based compression, whereas wildcard fields are whole-value LZ4 compression.
6. Will vary with content, but a test indexing weblogs took 499 seconds vs. keyword's 365 seconds.

### Decision Flow

Since deciding between `wildcard` and `keyword` involves weighing trade-offs, this workflow is a visual to help assess when choosing `wildcard` may provide an advantage [2].

<img width="929" alt="wildcard-field-workflow" src="https://images.contentstack.io/v3/assets/bltefdd0b53724fa2ce/blt086af7a2897168a8/5f37050cdb5c28785b6f0413/blog-wildcard-field-workflow.png">

### Use Cases

The following sections detail use cases which could benefit using the `wildcard` type.

#### Stack traces

> Program stack traces tend to be well-structured but with long text and varied contents. There are too many subtleties and application-specific patterns to map all of them accurately with ECS' field definitions. Better performing wildcard searches can help users formulate their queries easier and with a more minor performance hit.

Reviewer: ```suggestion
Program stack traces tend to be well-structured but with long text and varied contents. There are too many subtleties and application-specific patterns to map all of them accurately with ECS' field definitions. Better performing wildcard searches can help users formulate their queries easier and with a less significant performance hit.
```

1. n-gram index of all three character sequences in the string values
2. binary doc value store of the original values

Wildcard fields require more disk space for the additional n-gram index. This disk cost is only recovered if the full doc values are also stored compress better than the compression of the keyword fields for the doc values. Keyword values are compressed based on common-prefixes in the values, but wildcard values use LZ4 based on blocks of 32 values. These two compression approaches vary depending on size, duplicate values, the cardinality of the data, and so on.

### Storage and Indexing Costs

> When assembling the initial list of candidate fields to migrate to `wildcard`, we split focus between query performance improvements and removing security blind spots. However, we overlooked the storage and indexing costs when switching fields to be indexed as `wildcard.`

Reviewer: ```suggestion
When assembling the initial list of candidate fields to migrate to `wildcard`, we split focus between query performance improvements and removing security blind spots. However, we overlooked the storage and indexing costs when switching fields to be indexed as `wildcard`.
```

### Ingestion

> Any component producing data (Beats, Logstash, third-party developed, etc.) will need to adopt the mappings in their index templates.

Reviewer: up to you
```suggestion
Any component producing data (Agent, Beats, Logstash, third-party developed, etc.) will need to adopt the mappings in their index templates.
```

### Wildcard and case-insensitivity

> Some fields require flexibility in how users search. For example, their content is messy (such as user-agent values) or popular for threat hunters (command line processes), and a single character in the opposite casing can bypass a detection today for `keyword` fields. The `wildcard` field provides improved performance of leading `wildcard` and `regex` term-level queries but is also a step towards case-insensitive search support in Elasticsearch.

Reviewer: Both types support case insensitive search, since 7.10 IIRC. Might want to revisit this sentence

```suggestion
Some fields require flexibility in how users search. For example, their content is messy (such as user-agent values) or popular for threat hunters (process command-lines or a PowerShell script). The `wildcard` field provides improved performance of leading `wildcard` and `regex` term-level queries.
```

Reviewer: We use EQL for all of our security rules that require explicit case insensitivity, because it makes it easy to go back and forth

Author: Thanks - will revisit!

#### Resolution

The `case_insensitivity` query parameter was added in Elasticsearch 7.10. Both `keyword` and `wildcard` types are supported, and each type's noted [performance characteristics](#comparison-with-keyword) will be consistent.

### Performance differences

> Performance and storage characteristics between wildcard and keyword will be different[4]. This difference will vary based deployment size and/or the amount of field data duplication. Fields that were previously indexed as keyword will switch to wildcard. With these fields now indexed as wildcard, users will be querying fields that are indexed as keyword in some indices and as wildcard in others. Any possible indexing or querying differences need to be understood and captured.

Reviewer: ```suggestion
Performance and storage characteristics between `wildcard` and `keyword` will be different[4]. This difference will vary based deployment size and/or the amount of field data duplication. Fields that were previously indexed as `keyword` will switch to `wildcard`. With these fields now indexed as `wildcard`, users will query fields that are indexed as keyword in some indices and as wildcard in others. Any possible indexing or querying differences need to be understood and captured.
```

Indexing and query performance characteristics of both types were explored. The observations were [noted](#comparison-with-keyword) earlier in this proposal. However, after additional benchmarking, the increases in storage costs and decreasing in index performance were found to be significant enough that we need to revisit our approach.

#### Resolution

The following categories were initially candidates for `wildcard`, but after reviewing the benchmarking data, the fields will not typically have high enough cardinality to make them ideal candidates for `wildcard`.
##### File paths and names

> File path values are likely to compress well as `keyword` since `keyword` fields have common-prefix-based compression (`wildcard` values are blocks of 32 values compressed into a single LZ4 blob). On top of the worse doc values compression, the number of `postings` also increases significantly due to n-grams.

Reviewer: ```suggestion
File path values are likely to compress well as `keyword` since `keyword` fields have common-prefix-based compression (`wildcard` values are blocks of 32 values compressed into a single LZ4 blob). In addition to worsened doc values compression, the number of `postings` also increases significantly due to n-grams.
```

##### Host and Organization Naming

> Often, hostname values duplicate from event to event. An index will usually have thousands of different hosts and is unlikely to see millions of unique hostname values.

Reviewer: Slightly more clear to me, because I can more easily tell which part of speech "duplicate" is
```suggestion
Often, hostname values are duplicated from event to event. An index will usually have thousands of different hosts and is unlikely to see millions of unique hostname values.
```

For certain field and their values, the use of specialized text analysis could be an alternative to using the `wildcard` data type. For example, adopting the [path hierarchy tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-pathhierarchy-tokenizer.html) for file paths.

### Wildcard field value character limits

ECS applies the `ignore_above` setting to keyword fields to prevent strings longer than 1024 characters from being indexed or stored. Users may raise the setting of `ignore_above`, but Lucene implements a term byte-length limit of 32766, which cannot be adjusted. Wildcard supports an unlimited max character size for a field value. The `wildcard` field type will still have the `ignore_above` option available, and a reasonable limit may need applying to mitigate unexpected side-effects.

#### Resolution

> This ability to ingest very long values is considered an advantage of `wildcard` compared to `keyword`. Therefore, the `wildcard` fields will not have an `ignore_above` option defined initially.

Reviewer: 🤎
